src/Plugin/tmgmt/Translator/ThebigwordTranslator.php

```diff
@@ -0,0 +1,872 @@
+   *   If we expect resource to be downloaded.
+   *
+   * @return array
+   *   Response array from Thebigword.
+   *
+   * @throws \Drupal\tmgmt\TMGMTException
+   */
+  public function request($path, $method = 'GET', $params = array(), $download = FALSE) {
+    $options = array();
+    if (!$this->translator) {
+      throw new TMGMTException('There is no Translation entity. Access to public/secret keys is not possible.');
+    }
+
+    if (\Drupal::config('tmgmt_thebigword.settings')->get('use_mock_service')) {
+      $url = $GLOBALS['base_url'] . '/tmgmt_thebigword_mock' . '/v' . self::API_VERSION . '/' . $path;
+    }
+    else {
+      $url = self::PRODUCTION_URL . '/v' . self::API_VERSION . '/' . $path;
+    }
+
+    try {
+      if ($method == 'GET') {
+        $options['query'] = $params;
+      }
+      else {
+        $options['json'] = $params;
+      }
+      $options['headers'] = [
+        'TMS-REQUESTER-ID' => $this->translator->getSetting('client_contact_key'),
+      ];
+      $response = $this->client->request($method, $url, $options);
+    }
+    catch (BadResponseException $e) {
+      $response = $e->getResponse();
+      // debug($response->getBody()->getContents());
+      throw new TMGMTException('Unable to connect to Thebigword service due to following error: @error', ['@error' => $response->getReasonPhrase()], $response->getStatusCode());
+    }
+
+    if ($response->getStatusCode() != 200) {
+      throw new TMGMTException('Unable to connect to the Thebigword service due to following error: @error at @url',
+        array('@error' => $response->getStatusCode(), '@url' => $url));
+    }
+
+    // If we are expecting a download, just return received data.
+    $received_data = $response->getBody()->getContents();
+    if ($download) {
+      return $received_data;
+    }
+    $received_data = json_decode($received_data, TRUE);
+
+    return $received_data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSupportedRemoteLanguages(TranslatorInterface $translator) {
+    if (!empty($this->supportedRemoteLanguages)) {
+      return $this->supportedRemoteLanguages;
+    }
+
+    try {
+      $this->setTranslator($translator);
+      $supported_languages = $this->request('languages', 'GET', array());
+
+      // Parse languages.
+      foreach ($supported_languages as $language) {
+        $this->supportedRemoteLanguages[$language['CultureName']] = $language['DisplayName'];
+      }
+
+      // In case of failed request or parsing, we are returning a list of
+      // supported remote languages from default Thebigword mapping.
+      if (empty($this->supportedRemoteLanguages)) {
+        return [];
+      }
+    }
+    catch (\Exception $e) {
+      if (\Drupal::routeMatch()->getRouteName() == 'entity.tmgmt_translator.edit_form'
+        || \Drupal::routeMatch()->getRouteName() == 'entity.tmgmt_translator.add_form') {
+        return [];
+      }
+      drupal_set_message($e->getMessage(), 'error');
```

Comment:
I'm skipping an error the first time that you edit the translator and there is no client contact key yet.